### PR TITLE
generic fetcher: Generate SBOM components

### PR DESCRIPTION
**Pull request by kosciCZ**
Generic fetcher now generates SBOM components of purl type pkg:generic. It also produces externalReferences of type distribution for each generic component to allow for gating at later time.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

File: cachi2/core/models/sbom.py
Comment (slimreaper35) [inline cachi2/core/models/sbom.py:1]: The changes in this file would deserve a separate commit.

File: tests/unit/package_managers/test_generic.py
Comment (slimreaper35) [inline tests/unit/package_managers/test_generic.py:211]:
```suggestion
    ["lockfile_content", "expected_components"]
```
nitpick: I would choose slightly more verbose names.

File: cachi2/core/package_managers/generic/main.py
Comment (slimreaper35) [inline cachi2/core/package_managers/generic/main.py:125]: Is `package_manager` an attribute in the `Component` class ?

File: cachi2/core/models/sbom.py
Comment (kosciCZ) [reply cachi2/core/models/sbom.py:1]: Split to a separate commit.

File: tests/unit/package_managers/test_generic.py
Comment (kosciCZ) [reply tests/unit/package_managers/test_generic.py]: Changed.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: Good catch, that's a leftover from something I was testing. Removed.

File: cachi2/core/package_managers/generic/main.py
Comment (slimreaper35) [inline cachi2/core/package_managers/generic/main.py:122]: Maybe one more suggestion, `a` and `h` do not tell much about what it stands for.  Btw. in the `LockfileArtifact` you have used: `ChecksumInfo(algo, digest) for algo, digest in self.checksums.items()`. For readability, you can define checksums just below the `url` variable.

But, I will approve the PR.

**Review by a-ovchinnikov — APPROVED** [review]
LGTM

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py]: I agree, it dows improve readability. Fixed.

**Comment by kosciCZ** [comment]
This has acquired the necessary number of approvals. Can this be merged? It'd be great if this made it to this week's release.

File: cachi2/core/models/sbom.py
Comment (eskultety) [inline cachi2/core/models/sbom.py:30]: Can you justify this change? The commit message is empty with such an IMO substantial SBOM update. Why is this needed? Why have these specific fields been chosen for the external reference? Can you elaborate?

**Comment by eskultety** [comment]
Please don't merge until https://github.com/containerbuildsystem/cachi2/pull/710#discussion_r1820475942 is addressed.

File: cachi2/core/models/sbom.py
Comment (kosciCZ) [reply cachi2/core/models/sbom.py:30]: It is sort of described in the follow-up commit message, as I'd say the introduction itself bears no meaning, if not used within the context of the generic fetcher (other package managers do not utilize it).

> Generic fetcher now generates SBOM components of purl type
pkg:generic. It also produces externalReferences of type
distribution for each generic component to allow for gating
at a later time.

I chose `externalReferences` of type `distribution` to describe the `download_url` in a way that's easily processable in whatever will look at the SBOM later (thinking about EC, but could be anything, really). The [CycloneDX spec](https://github.com/CycloneDX/guides/blob/main/SBOM/en/0x51-External-References.md#external-references) describes this type of externalReferences as "Direct or repository download location" which in my opinion makes sense for the `download_url`

File: cachi2/core/package_managers/generic/main.py
Comment (brunoapimentel) [inline cachi2/core/package_managers/generic/main.py:115]: We're reporting all checksums present in the lockfile here, but this does not mean that all of them were validated. I'm wondering if this is the behavior we want. It'd be nice to double check how this is being done across different package managers, but I believe we can investigate this as a follow up, no need to block this PR.

File: cachi2/core/models/sbom.py
Comment (brunoapimentel) [reply cachi2/core/models/sbom.py:30]: I think what you wrote here can be worked into a message for this specific commit. For now, it'll only be used for `generic`, but that might change in the future. The reasoning here is that this field should carry the URL from which a component was downloaded, as a means of having more meaningful data in the SBOM so that other tooling can build on top of it.

File: cachi2/core/package_managers/generic/main.py
Comment (kosciCZ) [reply cachi2/core/package_managers/generic/main.py:115]: Huh, I guess that's technically right. I guess we would need to implement something like `must_match_all_checksums`? Good point though.

File: cachi2/core/models/sbom.py
Comment (kosciCZ) [reply cachi2/core/models/sbom.py:30]: I will amend the commit message for that commit.

Comment (kosciCZ) [reply cachi2/core/models/sbom.py:30]: Amended.

File: cachi2/core/package_managers/generic/main.py
Comment (brunoapimentel) [reply cachi2/core/package_managers/generic/main.py:115]: From a quick look, it seems that the currently existing package managers are only reporting a single checksum in the SBOM. So I think it's worth to have a discussion about this before we end the experimental phase on this.

**Review by brunoapimentel — APPROVED** [review]

File: cachi2/core/models/sbom.py
Comment (eskultety) [reply cachi2/core/models/sbom.py:30]: @kosciCZ  please include https://cyclonedx.org/docs/1.6/json/#components_items_externalReferences as a reference in the commit message where the following is IMO the most important bit justifying the selection of this particular component item:

> External references provide a way to document systems, sites, and information that may be relevant but are not included with the BOM. They may also establish specific relationships within or external to the BOM.

File: cachi2/core/package_managers/generic/main.py
Comment (eskultety) [reply cachi2/core/package_managers/generic/main.py:115]: > From a quick look, it seems that the currently existing package managers are only reporting a single checksum in the SBOM. So I think it's worth to have a discussion about this (should we report checksums that were not validated) before we end the experimental phase on generic.

@brunoapimentel can you please open an issue on this so that that discussion is not forgotten?

**Review by eskultety — APPROVED** [review]
@kosciCZ with the CycloneDX reference added to commit nr.1 -> ACK

File: cachi2/core/models/sbom.py
Comment (kosciCZ) [reply cachi2/core/models/sbom.py:30]: Added to the commit message.

**Pull request merged**
